src/app/lib/pagination.ts

```diff
@@ -7,33 +7,73 @@ type PaginatedResult<T> = {
   data: T;
 };
 
+export type PaginatedResultEx<T> = {
+  meta: {
+    has_next: boolean;
+    total_returned: number;
+    next_offset: number;
+  };
+  data: T;
+};
+
+export type PaginationParams = {
+  page: number;
+  pageSize: number;
+};
+
+export type PaginationParamsEx = {
+  limit: number;
+  offset: number;
+};
+
 export async function paginateResult<T extends Array<any>>(
-  result: (skip: number, take: number) => Promise<T>,
+  query: (skip: number, take: number) => Promise<T>,
   page: number,
   pageSize: number
 ): Promise<PaginatedResult<T>> {
   const skip = (page - 1) * pageSize;
   const take = pageSize + 1;
 
-  const data = await result(skip, take);
+  const ex = await paginateResultEx<T>(query, {limit: take, offset: skip});
+
+  return {
+    meta: {
+      currentPage: page,
+      pageSize: pageSize,
+      hasNextPage: ex.meta.has_next,
+    },
+    data: ex.data,
+  };
+}
+
+export async function paginateResultEx<T extends Array<any>>(
+  query: (skip: number, take: number) => Promise<T>,
+  params: PaginationParamsEx
+): Promise<PaginatedResultEx<T>> {
+  // retrieve one more than requested to see if there is more data
+  // for user to query
+  const data = await query(params.offset, params.limit + 1);
 
   if (!data || data.length === 0) {
-    // return notFound();
     return {
-      meta: { currentPage: 0, pageSize: 0, hasNextPage: false },
+      meta: { 
+        has_next: false,
+        total_returned: 0, 
+        next_offset: 0
+      },
       data: [] as any as T,
     };
   }
 
-  const hasNextPage = data.length > pageSize;
-  const theData = data.slice(0, pageSize) as T;
+  const has_next = data.length > params.limit;
+  const theData = data.slice(0, params.limit) as T;
 
   return {
     meta: {
-      currentPage: page,
-      pageSize,
-      hasNextPage,
+      has_next: has_next,
```

Comment:
is there a reason to change the meta format from hasNextPage to has_next etc?

Reply:
The pagination semantics should not change for existing calls to PaginateResult. The semantics are slightly different for our external API (which is using limit/offset as opposed to pageSize + page index). 

The reason for limit / offset for our API is that it seems to be a slightly-more-often-used semantic compared to page size, so was using that to keep dev integration as simple as possible.

completely open to either snake/camel case for the pagination metadata. I really have no strong opinion either way, we just need to be consistent once we start exposing it.